source/slang/slang-check-modifier.cpp

```diff
@@ -628,13 +628,102 @@ namespace Slang
         else if (auto customJVPAttr = as<CustomJVPAttribute>(attr))
         {
             SLANG_ASSERT(attr->args.getCount() == 1);
-            
+            SLANG_ASSERT(as<Decl>(attrTarget));
+
             // Ensure that the argument is a reference to a function definition or declaration.
-            auto funcExpr = as<DeclRefExpr>(CheckTerm(attr->args[0]));
-            if (!as<FuncType>(funcExpr->type))
+            auto diffExpr = CheckTerm(attr->args[0]);
+            if (diffExpr->type == getASTBuilder()->getErrorType())
+            {
+                // Could not resolve the term.
+                getSink()->diagnose(diffExpr, Slang::Diagnostics::invalidCustomDerivative, as<Decl>(attrTarget));
+                return false;
+            }
+
+            // Either diffExpr has a function type, or it is a reference to a generic.
+            if (!as<FuncType>(diffExpr->type) && 
+                    !(as<DeclRefExpr>(diffExpr) && 
+                      as<DeclRefExpr>(diffExpr)->declRef.as<GenericDecl>().getDecl() != nullptr))
+            {
                 return false;
+            }
 
-            customJVPAttr->funcDeclRef = funcExpr;
+            auto diffDeclRef = as<DeclRefExpr>(diffExpr)->declRef;
+
+            UCount genericLevels = 0;
+            // If we've grabbed the outer generic for some reason,
+            // recursively construct GenericAppExpr<...>(generic)
+            // and check that to get a specialized func.
+            // 
+            while (diffDeclRef.as<GenericDecl>().getDecl() != nullptr)
```

Comment:
I think this is not necessary. The `diffDeclRef` already contains subsitution info for the generic args. Just create a `DeclRefExpr` with `diffDeclRef` and use it  I guess the missing part is that we can't easily fill in the `type` field for the declref expr without invoking `CheckTerm`, so perhaps consider making `customJVPAttr->funcDeclRef` just a declref?

Reply:
Yes, I realized about 80% through the implementation that I could use `__custom_jvp(diff_fn<T>)` instead of using this logic, but I decided to commit it anyway since we can also use just `__custom_jvp(diff_fn)` and have the substitutions made automatically.

We can remove it though, it might make things break later on.